go-update/src/main/java/org/reactome/release/goupdate/DuplicateCleaner.java

```diff
@@ -0,0 +1,182 @@
+			List<Long> dbIDsWithNoReferrers = new ArrayList<Long>();
+			Map <Long,Integer> refCounts = dupeReporter.getReferrerCountForAccession(accession, ReactomeJavaConstants.GO_BiologicalProcess, ReactomeJavaConstants.GO_MolecularFunction, ReactomeJavaConstants.GO_CellularComponent, ReactomeJavaConstants.Compartment);
+			// if there are referrers...
+			if (refCounts != null && refCounts.size() > 0)
+			{
+				// for each DB ID in referrer counts
+				for (Long dbId : refCounts.keySet())
+				{
+					// if there are referrers.
+					if (refCounts.get(dbId) > 0)
+					{
+						logger.info("\tAccession instance with DB_ID {} has {} significant referrers", dbId, refCounts.get(dbId));
+
+						// update count.
+						instancesWithSignificantReferrers ++;
+						GKInstance inst = adaptor.fetchInstance(dbId);
+						@SuppressWarnings("unchecked")
+						Collection<GKSchemaAttribute> refAttribs = (Collection<GKSchemaAttribute>) inst.getSchemClass().getReferers();
+						// check each atttribute
+						for (GKSchemaAttribute refAtt : refAttribs)
+						{
+							@SuppressWarnings("unchecked")
+							Collection<GKInstance> refs = (Collection<GKInstance>) inst.getReferers(refAtt);
+							if (refs != null && refs.size() > 0)
+							{
+								for (GKInstance ref : refs)
+								{
+									if (!Arrays.asList(ReactomeJavaConstants.GO_BiologicalProcess, ReactomeJavaConstants.GO_MolecularFunction, ReactomeJavaConstants.GO_CellularComponent, ReactomeJavaConstants.Compartment).contains(ref.getSchemClass().getName()))
+									{
+										logger.info("\t\tvia {}:\t{}", refAtt.getName(), ref.toString());
+									}
+								}
+							}
+						}
+					}
+					// if there are NO referrers, add the db_id to the list of DB IDs without referrers.
+					else
+					{
+						logger.debug("DB ID {} for accession {} has no significant referrers, and will be deleted.", dbId, accession);
+						dbIDsWithNoReferrers.add(dbId);
+					}
+				}
+			}
+			// update list of things to delete with IDs that have no referrers.
+			dbIDsToDelete.addAll(dbIDsWithNoReferrers);
+			// If no DB IDs have significant referrers, we should delete ALL DB_IDs for the accession, except for the newest one.
+			if (dbIDsWithNoReferrers.size() == refCounts.keySet().size())
+			{
+				GKInstance newestInstance = null;
+				LocalDateTime newestDate = null;
+				DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.S");
+				// Need to try all classes since we don't know which class a GO Accession might be.
+				for (String reactomeClass : Arrays.asList(ReactomeJavaConstants.GO_BiologicalProcess, ReactomeJavaConstants.GO_MolecularFunction, ReactomeJavaConstants.GO_CellularComponent))
+				{
+					GKInstance modifiedInstance = null;
+					
+					@SuppressWarnings("unchecked")
+					Collection<GKInstance> instances = (Collection<GKInstance>) adaptor.fetchInstanceByAttribute(reactomeClass, ReactomeJavaConstants.accession, "=", accession);
```

Comment:
The for loop to look at the different GO classes could be done in a method so the detail of grabbing GO instances is extracted away and the nesting is reduced.  Something like:

```
private List<GKInstance> getGOInstancesByAccession(MySQLAdaptor adaptor, String accession) throws Exception {
	List<GKInstance> GOInstances = new ArrayList<>();
	for (String reactomeClass : Arrays.asList(
		ReactomeJavaConstants.GO_BiologicalProcess,
		ReactomeJavaConstants.GO_MolecularFunction,
		ReactomeJavaConstants.GO_CellularComponent)
	) {

	    @SuppressWarnings("unchecked")
	    Collection<GKInstance> instances = (Collection<GKInstance>) adaptor.fetchInstanceByAttribute(reactomeClass, ReactomeJavaConstants.accession, "=", accession);
            if (GOInstances != null) {	    
                GOInstances.addAll(instances);
            }
	}
	return GOInstances;
}
```